single_kernel_mongo/events/ldap.py

```diff
@@ -0,0 +1,118 @@
+        self.framework.observe(
+            self.manager.ldap_requirer.on.ldap_unavailable, self._on_ldap_unavailable
+        )
+        self.framework.observe(
+            self.manager.certificate_transfer.on.certificate_available,
+            self._on_certificate_available,
+        )
+        self.framework.observe(
+            self.manager.certificate_transfer.on.certificate_removed, self._on_certificate_removed
+        )
+        self.framework.observe(self.restart_if_ready_event, self._on_restart_if_ready)
+
+        self.framework.observe(
+            self.charm.on["ldap-peers"].relation_changed, self._on_restart_if_ready
+        )
+
+    def _on_ldap_ready(self, event: LdapReadyEvent):
+        """Handles the ops event that indicates that ldap relation is ready."""
+        action = "ldap-ready"
+        try:
+            self.manager.on_ldap_ready(event.relation)
+        except WaitingForLdapDataError as err:
+            self.charm.status_manager.set_and_share_status(WaitingStatus("Waiting for LDAP data."))
+            defer_event_with_info_log(logger, event, action, f"{err}")
+
+        except (DeferrableError, DeferrableFailedHookChecksError) as err:
+            defer_event_with_info_log(logger, event, action, f"{err}")
+        except LDAPSNotEnabledError:
+            self.charm.status_manager.set_and_share_status(
+                BlockedStatus("LDAPS not enabled on LDAP application.")
+            )
+        except NonDeferrableFailedHookChecksError as err:
+            logger.error(f"{err}")
+            self.charm.status_manager.set_and_share_status(BlockedStatus(err.args[0]))
+
+    def _on_ldap_unavailable(self, event: LdapUnavailableEvent):
+        """Handles the ops event that indicates that ldap relation is now unavailable."""
+        self.manager.on_ldap_unavailable()
+
+    def _on_certificate_available(self, event: CertificateAvailableEvent):
+        """Handles the ops event that indicates that ldap-certificates relation is ready."""
+        try:
+            self.manager.on_certificate_available(event.certificate, event.ca, event.chain)
+        except DeferrableFailedHookChecksError as err:
+            defer_event_with_info_log(logger, event, "ldap-cert-ready", f"{err}")
+        except NonDeferrableFailedHookChecksError as err:
+            logger.error(f"{err}")
+            self.charm.status_manager.set_and_share_status(BlockedStatus(err.args[0]))
+
+    def _on_certificate_removed(self, event: CertificateRemovedEvent):
+        """Handles the ops event that indicates that ldap-certificates relation is unavailable."""
+        self.manager.on_certificate_removed()
+
+    def _on_restart_if_ready(self, event: RestartIfReadyEvent):
+        """Custom ops revent to trigger restart of leader with a single source of truth."""
+        action = "restart-ldap-if-ready"
+        try:
+            self.manager.restart_when_ready()
```

Comment:
There seems to be a bit of a discrepancy between the manager restarts and this one, I see 3 points of entry for a restart:
- restart_charm_services 
- restart_when_ready
- restart_if_ready_event.emit()

Reply:
All restarts go trough the `restart_when_ready` which  first runs some checks and then calls the `restart_charm_services`.

`restart_when_ready` is called by the relation changed events (for the non leader units) or by the custom ops event (for the leader unit).